### PR TITLE
EZP-30503: Fixed missing default value for the new page_layout parameter

### PR DESCRIPTION
**Pull request by emodric**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30503](https://jira.ez.no/browse/EZP-30503)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | no

https://github.com/ezsystems/ezpublish-kernel/pull/2629 was incomplete. Default value for `page_layout` parameter did not exist.

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

**Comment by emodric** [comment]
Ping @alongosz

**Comment by emodric** [comment]
@micszo Can you share the details of your test so I can reproduce? I don't see a way this patch can cause the issue you describe.

**Comment by micszo** [comment]
In `/templates` add `template.html.twig` (e.g. with `<p>Lorem ipsum.</p>`). 
In `config/packages/ezplatform.yaml` (with two siteaccesses: site, fre) add at first `ezpublish.system.default.site.pagelayout: 'template.html.twig'` and then try `ezpublish.system.default.site.page_layout: 'template.html.twig'`
Reload cache.
Try loading frontend on both siteaccesses: _hostname/site_, _hostname/fre_.
The _/fre_ siteaccess always works as usual, result for _/site_ depends on using `pagelayout` or `page_layout`. 
And it's the opposite result depending whether this branch is used or not.

**Comment by alongosz** [comment]
@emodric I planned to look at this at some point, but if you have time I appreciate any help with this (going through a backlog of other tickets ;) )

**Comment by emodric** [comment]
@micszo `ezpublish.system.default.site.page_layout` format of the parameter looks wrong.

Should it be maybe like this?

```
ezpublish:
    system:
        default:
            pagelayout: 'template.html.twig'
            page_layout: 'template.html.twig'
```

And then you should be able to use this parameter in a template with `ezplatform.configResolver.parameter('pagelayout')` or `ezplatform.configResolver.parameter('page_layout')`.

**Comment by micszo** [comment]
My bad, indeed I had `ezpublish.system.site.page_layout` (`site` being at the same level as `default`).

**Comment by micszo** [comment]
So will only `default` work?

**Comment by emodric** [comment]
So the patch works fine? :)

**Comment by emodric** [comment]
> So will only `default` work?

It doesn't matter. Any siteaccess or siteaccess should work. `default` scope is used only to make sure that the parameter is available on every siteaccess.

**Comment by micszo** [comment]
I used `site`. The mistake was only in the comment here, in `ezplatform.yaml` I had `ezpublish.system.site.page_layout`.

**Comment by emodric** [comment]
After this patch, and with this config:

```
ezpublish:
    system:
        site:
            pagelayout: 'template.html.twig'
            page_layout: 'template.html.twig'
```

You should be able to access both parameters: `{{ dump(ezplatform.configResolver.parameter('pagelayout')) }}` and `{{ dump(ezplatform.configResolver.parameter('page_layout')) }}`

**Comment by emodric** [comment]
Also, I have tested the patch again and it works as expected. Once you configure the pagelayout as shown, the parameter is available in config resolver (and previously it wasn't).

**Comment by micszo** [comment]
Have you also checked if it works on frontend for you? :)

**Comment by alongosz** [comment]
> Also, I have tested the patch again and it works as expected. Once you configure the pagelayout as shown, the parameter is available in config resolver (and previously it wasn't).

That's true, but the parameter is passed to e.g. [the default full view](https://github.com/ezsystems/ezpublish-kernel/blob/v8.0.0-beta1/eZ/Bundle/EzPublishCoreBundle/Resources/views/default/content/full.html.twig#L1). Semantic config for SA is the way to configure it, so if someone changes the config mentioned by @micszo, he should see his layout used for the given SiteAccess w/o trying to get it from ConfigResolver in e.g. custom controller. With the default templates used, setting `page_layout` or `pagelayout` should be enough.

At first glance, either logic of `eZ\Publish\Core\MVC\Symfony\View\ParametersInjector\ViewbaseLayout::getPageLayout` needs to be adjusted for BC magic to work, or this fix itself setting proper context values, keeping BC. I'd need to dig into this deeper.

This is btw. the reason we didn't want to handle BC on Twig variables, but to rename them only. There are always undiscovered issues which consume a lot of time to achieve something small :-)

**Comment by emodric** [comment]
For me, `{% extends page_layout %}` also works after my patch without issues.

**Comment by alongosz** [comment]
> For me, `{% extends page_layout %}` also works after my patch without issues.

Hmm, ok. Then I need to look into this, because for us it works as @micszo described.

**Comment by emodric** [comment]
Updated the comment, sorry. To reiterate: After the patch, for me both `{% extends page_layout %}` and `{% extends pagelayout %}` work in a full view.

**Comment by alongosz** [comment]
> Updated the comment, sorry. To reiterate: After the patch, for me both `{% extends page_layout %}` and `{% extends pagelayout %}` work in a full view.

If you don't write `{% extends ...` and rely on mentioned default view, then it should work as well. The idea behind this was that indeed both `pagelayout` and `page_layout` Twig variables should work, but also if you configure via SA configuration either `pagelayout` and `page_layout` setting, these Twig variables should be set properly, with `page_layout` taking precedence if both were configured (though probably doesn't make sense to have them both).

In other words, if you configure `pagelayout` SA setting, both `{% extends pagelayout %}` and `{% extends page_layout %}` should give that setting directly. The same goes for `page_layout` SA setting.

For us this isn't the case.

**Comment by emodric** [comment]
> If you don't write {% extends ... and rely on mentioned default view, then it should work as well.

I'm not sure I understand. What is the default view? For these parameters, you either use them by extending from them in full views, or get them from config resolver.

And yes, I think that this patch uncovers the issue in `eZ\Publish\Core\MVC\Symfony\View\ParametersInjector\ViewbaseLayout::getPageLayout` you mentioned, because the fallback when only `pagelayout` config is defined does not work since now `page_layout` param is always available, as it should be (and wasn't previously). So this patch is technically correct.

I agree however it should be merged once `eZ\Publish\Core\MVC\Symfony\View\ParametersInjector\ViewbaseLayout::getPageLayout` is "fixed".

> This is btw. the reason we didn't want to handle BC on Twig variables, but to rename them only. There are always undiscovered issues which consume a lot of time to achieve something small :-)

It is your job as a maintainer of a large scale project to take care of BC, however small it may be, right? I still think that what was done with all BC breaks in v8 should've been handled better. Sure, it makes your job easier, but it antagonizes everyone else. Project and 3rd party bundles/lib developers.

**Comment by alongosz** [comment]
> I'm not sure I understand. What is the default view? For these parameters, you either use them by extending from them in full views, or get them from config resolver.

[This one](https://github.com/ezsystems/ezpublish-kernel/blob/v8.0.0-beta1/eZ/Bundle/EzPublishCoreBundle/Resources/views/default/content/full.html.twig#L1). It works out of the box on clean install, w/o creating template for it yourself. Though in custom full views the behavior should be the same.

> I agree however it should be merged once `eZ\Publish\Core\MVC\Symfony\View\ParametersInjector\ViewbaseLayout::getPageLayout` is "fixed".

Can I push the fix (once I have it) to this PR? :) It's kind of related, indirectly.

> It is your job as a maintainer of a large scale project to take care of BC, however small it may be, right? I still think that what was done with all BC breaks in v8 should've been handled better. Sure, it makes your job easier, but it antagonizes everyone else. Project and 3rd party bundles/lib developers.

It's all about hours in a day and days in the week, and resources associated with that :)

**Comment by emodric** [comment]
> This one. It works out of the box on clean install, w/o creating template for it yourself. Though in custom full views the behavior should be the same.

Truth to be told, I have no idea what that extends does or what `view_base_layout` variable is. I've never seen it used in the wild before. Either you extend the pagelayout in full views, or don't extend it in all other views. I have never seen any other combination before :D

> Can I push the fix (once I have it) to this PR? :) It's kind of related, indirectly.

Sure.

> It's all about hours in a day and days in the week, and resources associated with that :)

Sure. I'm not dissatisfied with you personally. I (and others) am/are dissatisified with your management and their decision on how to handle it.

**Comment by alongosz** [comment]
> The test result is as follows:
> 
> * without this fix `pagelayout` works (template is loaded) but `page_layout` doesn't work (template is not loaded)
> * with the fix `page_layout` works but `pagelayout` doesn't work

@micszo Fixed via 247e22d. Right now both `pagelayout` and `page_layout` should work, with `page_layout` taking precedence if both set. BC was moved from `ViewbaseLayout` to SA Configuration Parser so it exists on the lowest level.

**Comment by alongosz** [comment]
> Sure. I'm not dissatisfied with you personally. I (and others) am/are dissatisified with your management and their decision on how to handle it.

To be fair, this was our entire Engineering team decision. We had ~15 year old backlog of stale code. We knew we didn't want to continue creating technical debt in the new major version. We also knew that we didn't have enough resources (measured by e.g. each Sprint's capacity) to handle both deprecation and delivering new eZ Platform 3.0 experience on Symfony 4, migrating to which will require development cost on your side anyway. 

In other words, focusing on keeping old code in the next major version drags us away from important 3.0 work and, as the issue solved by this PR shows, it is never as simple as adding just one line of code.

We identified parts of the system we want to be cautious about based on a cost for 3rd party Developers to adjust. Aligning with unified Twig template filters, functions and variables requires running simple rename refactoring, so it's not that complicated, comparing to e.g. Symfony 4 migration. 

If we were able to foresee the future before releasing v2.5.0, we of course would have provided deprecation and forward compatibility layer to be dropped in 3.0. However, in contrast to e.g. Symfony, we were not able to work on two versions of the Product at the same time.

We also estimate that the next chance for code cleanup will be in ~2 years, so for us it's really big maintenance issue to introduce deprecated code right now, in 3.0. 

That said, when doing the next cleanup, we will look more carefully at deprecating before dropping. It's just that this release is special due to a lot of changes we're forced to do because of Symfony 4 upgrade.

**Comment by emodric** [comment]
> Aligning with unified Twig template filters, functions and variables requires running simple rename refactoring, so it's not that complicated, comparing to e.g. Symfony 4 migration.

All of these are problematic, because while they are indeed small changes and are relatively simple to replace, there is no way for a 3rd party lib to support both 2.5 **LTS** and 3.0 in the same version which puts an additional strain on maintainers to support them and causes unnecessary issues and mistakes that are due to come becuase of it.

And this is especially true for Twig function names and signal slots which are two points that would've benefited most from a BC layer. BC layer could've been in a separate package disconnected from the main repo.

**Comment by alongosz** [comment]
> All of these are problematic, because while they are indeed small changes and are relatively simple to replace, there is no way for a 3rd party lib to support both 2.5 **LTS** and 3.0

Is it really the only blocker for setting Symfony dependency as `^3.4 || ^4.3`? Because I think, that by definition, due to Symfony changes this assumption of cross major version support is not possible. This is why we have major versions. By the book handling of that in 3rd party package should really be done by releasing packages' major version. In my experience trying to support everything in all versions is very error prone. Not to mention difficult to maintain (yes, for me it would be more difficult to maintain one version of my package which supports two major versions of a dependency, than supporting two separate versions of my package).

> BC layer could've been in a separate package disconnected from the main repo.

We thought about it, but I really cannot promise anything due to time constraints :(

Back to the actual topic, thanks @micszo for testing and thanks @emodric for the fix. Merging :tada:

**Comment by emodric** [comment]
As long as a package does not have any deprecations in Symfony 3.4, it should work all up to Symfony 4.4. So supporting a major Symfony version with a previous LTS is trivial and is certainly as simple as `^3.4 || ^4.3`. In my opinion, I would expect any package version to fully support at least one LTS version of Symfony and with the way Symfony does it, it is certainly possible. I can't say the same for eZ Platform, which is the whole issue here.

**Comment by gggeek** [comment]
"Because I think, that by definition, due to Symfony changes this assumption of cross major version support is not possible" => Did you investigate this or is it just a guestimate?
Also, it seems that Sf is doing a better job than eZ at deprecating stuff for one full major-version cycle before killing it.

With that said, I'll stop pestering you about BC. Thx for fixing this issue.

**Comment by emodric** [comment]
> With that said, I'll stop pestering you about BC. Thx for fixing this issue.

Me too :) Thanks from me too for the merge!
